Feature: Add Google Cloud PubSub Instrumentation

manojpandey opened the pull request:
- [x] Add span tags for gcps
- [x] Instrumentation for Pub/Sub
- [x] Add simple tests
- [x] ~Mock wherever necessary~
- [x] Add emulator for testing & update tests
- [x] Licence updates
- [x] Remove TODOs and prints

manojpandey commented:
@andrewslotin ready for review - CI will pass after https://github.com/instana/python-sensor/pull/304 is merged - and then will rebase with master

andrewslotin commented on instana/instrumentation/google/cloud/pubsub.py:
```diff
@@ -0,0 +1,106 @@
+try:
+    from google.cloud import pubsub_v1
+
+    logger.debug('Instrumenting Google Cloud Pub/Sub')
```
Similar to the SQLAlchemy, we want to write this log line only after the instrumentation succeeded

manojpandey replied:
ACK

andrewslotin replied:
Still misplaced 😉

andrewslotin commented on instana/instrumentation/google/cloud/pubsub.py:
```diff
@@ -0,0 +1,106 @@
+            attr = message.attributes
+            if attr:
+                # trace continuity
+                headers = {
```
Why do we need a new dict here?

andrewslotin commented on instana/instrumentation/google/cloud/pubsub.py:
```diff
@@ -0,0 +1,106 @@
+            tracer.inject(scope.span.context, Format.TEXT_MAP, headers)
+
+            # update the metadata dict with instana trace attributes
+            kwargs.update(headers)
```
I believe this would inject the trace context into the named parameters dict instead of metadata, right?

manojpandey replied:
updating this

andrewslotin commented on instana/span.py:
```diff
@@ -262,6 +261,10 @@ def __init__(self, span, source, service_name, **kwargs):
             self.k = 1  # entry
 
+        # unify the span operation_name for gcps-producer and gcps-consumer
+        if "gcps" in self.data:
```
I believe an operation name would be more reliable option.

manojpandey replied:
I'll use `span.operation_name` then

andrewslotin commented on instana/span.py:
```diff
@@ -450,6 +459,12 @@ def _populate_exit_span_data(self, span):
+            self.data["gcps"]["op"] = span.tags.pop('gcps.op', None)
+            self.data["gcps"]["projid"] = span.tags.pop('gcps.projid', None)
+            self.data["gcps"]["top"] = span.tags.pop('gcps.top', None)
+            self.data["gcps"]["sub"] = span.tags.pop('gcps.sub', None)
```
Similar to the above, producers never send subscriptions.

manojpandey replied:
ACK

andrewslotin commented on tests/apps/flask_app/pubsub.py:
```diff
@@ -0,0 +1,100 @@
+# Use PubSub Emulator exposed at :8432 for local testing
+# os.environ["PUBSUB_EMULATOR_HOST"] = "localhost:8432"
+
+SERVICE_FILE = os.path.join(os.path.dirname(__file__), 'service-account-info.json')
```
A minor note, but would it be more handy to read the credentials path from env instead of hardcoding it?

manojpandey replied:
Yes, there is a better way to do it. Refactored the whole access mechanism and places everything inside a new test app folder - check the latest changes

andrewslotin commented on tests/apps/flask_app/service-account-info.json:
```diff
@@ -0,0 +1,2 @@
+# create a service account for Pub/Sub on Google Cloud Console and replace 
```
I'd rather suggest we don't commit this file to avoid leaking secrets by accident

manojpandey replied:
LGTM

andrewslotin commented on tests/apps/flask_app/pubsub.py:
```diff
@@ -0,0 +1,100 @@
+import os
```
The copyright header is missing 😉

manojpandey replied:
Ah ;) I'll add it

andrewslotin reviewed (changes requested):
Looks very good 👍 Apart from the comments I made, I'd suggest we keep the pure Flask app inside of `tests/apps/flask_app` and move the PubSub into a separate folder just to make navigation easier.

manojpandey commented on instana/instrumentation/google/cloud/pubsub.py:
```diff
@@ -0,0 +1,106 @@
+            return wrapped(*args, **kwargs)
+        else:
+            callback = args[1]
+            return wrapped(*[args[0], callback_with_instana], **kwargs)
```
This is a little tricky, as args is a tuple and tuples are immutable. I updated it in the new version of the commit - check it out

manojpandey commented on instana/span.py:
```diff
@@ -308,6 +311,12 @@ def _populate_entry_span_data(self, span):
+        elif span.operation_name == "gcps-consumer":
+            self.data["gcps"]["op"] = span.tags.pop('gcps.op', None)
+            self.data["gcps"]["projid"] = span.tags.pop('gcps.projid', None)
+            self.data["gcps"]["top"] = span.tags.pop('gcps.top', None)
```
ACK

manojpandey commented:
@andrewslotin I refactored the code a bit including the test app and addressed all of the review comments. 
Sending it again to you for another round of review.

andrewslotin commented on instana/instrumentation/google/cloud/pubsub.py:
```diff
@@ -0,0 +1,106 @@
+            callback = args[1]
+            args_list = list(args)
+            args_list[1] = callback_with_instana
+            args = tuple(args_list)
```
Would it work to do something like:
```python
subscription, callback, *args = args
```
and than
```python
args = (subscription, callback_with_instana, *args)
```
? This way you could use `subscription` and `callback` instead of `args[0]` in the callback wrapper.

manojpandey replied:
Thanks, this will work - much neater ;)

andrewslotin commented on instana/instrumentation/google/cloud/pubsub.py:
```diff
@@ -0,0 +1,106 @@
+            tracer.inject(scope.span.context, Format.TEXT_MAP, trace_continuity_headers)
+
+            # update the metadata dict with instana trace attributes
+            kwargs['trace_continuity_headers'] = json.dumps(trace_continuity_headers)
```
The trace context is expected to be present as key-value pairs inside of message attributes. If we serialize it to JSON, a Node consume for example would not pick it up.

manojpandey commented:
All changes are incorporated, ready for review @andrewslotin

andrewslotin reviewed (approved):
Good job 👍

manojpandey commented:
Closes #305
